include/deal.II/lac/linear_operator.h

```diff
@@ -0,0 +1,1175 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2010 - 2015 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef __deal2__linear_operator_h
+#define __deal2__linear_operator_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/lac/vector_memory.h>
+
+#ifdef DEAL_II_WITH_CXX11
+
+#include <functional>
+#include <array>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declarations:
+
+template <typename Number> class Vector;
+template <typename Number> class BlockVector;
+
+template <typename Range, typename Domain> class LinearOperator;
+
+template <typename Range = Vector<double>,
+          typename Domain = Range,
+          typename Exemplar,
+          typename Matrix>
+LinearOperator<Range, Domain> linop (const Exemplar &, const Matrix &);
```

Comment:
Not strongly at all.

Personally, I prefer a shorter variant - this was the reason for the initial choice <code>linop</code>. All in all, the following naming choices have to be made:
```
namespace - yes, no?
class LinearOperator
linop
transpose_linop
inverse_linop
identity_linop
block_linop
blockdiag_linop
```
Just make a comment on how to rename them.

But before proposing ```block_diagonal_linear_operator```, I'd prefer to have a separate namespace and just ```block_diagonal```.